### PR TITLE
Documentation Part I - General docu

### LINKED ISSUE #4: Create documentation on data schema
Generate a MKDocs page with explanations about the `data` schema, functionalities, etc. 

This is broken into parts:
- [x] General structure and basic information (#24)
- [ ] Explanation on ModelSystem, AtomsState
- [ ] Explanation on ModelMethod, NumericalSettings
- [ ] Explanation on Outputs, SCFOutputs, Variables, PhysicalProperties

### PR DESCRIPTION
**Pull request by JosePizarro3**
@ndaelman-hu @JFRudzinski 

I created some docu page, and lay the structure for explaining this repo. Please, check it when you can. I will keep working on the `ModelSystem` part of the docu, as well as the other sections on "How to use the Simulation schema" and "The normalize function".

**Sorry for pushing very hard these things**, but I have the feeling we need to sprint a bit to have some stuff done before May-June. I don't want to impose anything, tho.

I want to use this docu page to open more the discussion with PIs in Area C (so that they can actually get involved and give feedback), as I am sure they will be happy to give feedback to the documentation page. I have the feeling this could evolved into an ontology very easily, but for this @ndaelman-hu can give his thoughts.

### CODE REVIEW

**Comment by JosePizarro3** [comment]
Ah, you can checkout locally and do `mkdocs serve` as usual, to visualize the MKDocs page.

File: docs/general.md [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+
+We defined a base class named `Simulation`. You can find its Python schema definition in [src/nomad_simulations/general.py](https://github.com/nomad-coe/nomad-schema-plugin-simulation-data/blob/develop/src/nomad_simulations/general.py#L168). In NOMAD, this section will
+appear under the **_data_** section for the _archive_ of each _entry_.
+
+In NOMAD, a set of [base sections](https://nomad-lab.eu/prod/v1/staging/docs/howto/customization/base_sections.html) derived from the [Basic Formal Ontology (BFO)](https://basic-formal-ontology.org/) are defined. We used them to define `Simulation` as an [`Activity`](http://purl.obolibrary.org/obo/BFO_0000015). The UML diagram is:
+
+<div class="click-zoom">
+    <label>
+        <input type="checkbox">
+        <img src="../assets/simulation_base.drawio.png" alt="Simulation base section diagram." width="80%" title="Click to zoom in">
```
Comment (JFRudzinski) [inline docs/general.md]: What does the "s" in the diagrams following the type mean? I.e., `np.float64, s`? It is not immediately clear to me.

File: docs/general.md [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+The detailed UML diagram of quantities and functions defined for `Simulation` is:
+
+<div class="click-zoom">
+    <label>
+        <input type="checkbox">
+        <img src="../assets/simulation.drawio.png" alt="Simulation quantities and functions UML diagram." width="50%" title="Click to zoom in">
+    </label>
+</div>
+
+We use double inheritance from `EntryData` in order to populate the `data` section in the NOMAD archive. You can read more about the public normalize function function and its purpose in our base classes in the corresponding [documentation page](normalize.md). The private function `set_system_branch_depth()` is related with the `ModelSystem` base class, so we refer to its specific [documentation page](model_system/model_system.md) to know more.
```
Comment (JFRudzinski) [inline docs/general.md]: I would prefer using the name of the page for the link instead of the generic "documentation page":

"We use double inheritance from `EntryData` in order to populate the `data` section in the NOMAD archive. All of the base classes discussed here are subject to the [public normalize function](normalize.md) in NOMAD. The private function `set_system_branch_depth()` is related with the [ModelSystem base class](model_system/model_system.md)."

Check the statement about the normalize function, not sure if this is exactly what you meant, but just an example.

File: docs/general.md [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+4. `Outputs`: contains all the output properties, as well as references to the `ModelSystem` and `ModelMethod` used to obtain such properties. It might also contain information which will populate `ModelSystem` (e.g., atomic occupations, atomic moments, crystal field energies, etc.).
+
+!!! note "Self-consistent steps, SinglePoint entries, and more complex workflows."
+    In NOMAD, we consider the minimal unit for storing the data in the archive (i.e., an *entry*) as any calculation containing all the self-consistent steps of itself. This is what we call, `SinglePoint`. Thus, we do not split each self-consistent step in its own entry in the NOMAD archive, but rather store them under the same entry in the archive. Any other complex calculation which combines several differentiated self-consistent calculations must be considered a **workflow** (e.g., a `GW` calculation is usually composed of 2 `SinglePoint` entries: the `DFT SinglePoint` self-consistent calculation + the `GW SinglePoint` self-consistent calculations). You can check the corresponding [Github project](https://github.com/nomad-coe/nomad-schema-plugin-simulation-workflow) for more information.
```
Comment (JFRudzinski) [inline docs/general.md]: Similar to above, I would replace "GitHub project" with something like "NOMAD simulation workflow schema"

**Review by JFRudzinski — APPROVED** [review]
@JosePizarro3 Looks like a great start, thanks a lot for taking the initiative and getting this started! Just a few minor points to start, but not important and anyway this will all develop as we go, I think you should add this or even any other additions as soon as you would like.

File: docs/general.md (hunk 1 above)
Comment (JosePizarro3) [reply docs/general.md]: These are the units. Yeah, I agree it is pretty obscure. Should I remove them or come up with a more clear naming?

**Comment by JosePizarro3** [comment]
@JFRudzinski @Bernadette-Mohr I finished this part of the documentation, leaving empty the specific sub-parts when talking about `ModelSystem`, `ModelMethod`, `Outputs` and on "How to use the schema", as we need more work before being able to write something.

There is also an open question at the end of "The normalize function" section.

Maybe you can drop a review? @Bernadette-Mohr I can help you visualize the MKDocs page locally so you can have a better idea of the docu page, let me know.

File: docs/general.md (hunk 1 above)
Comment (JosePizarro3) [reply docs/general.md]: Added a note

File: docs/normalize.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+By checking on the `normalize()` functions and **rule 3**, we can establish whether `ArchiveSection.normalize()` will be ran or not. In `Section1.normalize()`, it will not, while in the other sections, `Section2` and `ParentSection`, it will.
+
+
+### Avoiding double execution of public functions
+
+When the `MetainfoNormalize` is ran, and the `normalize()` functions are executed, these usually have calls for calling private and public functions in the base classes which might have been used already during parsing the files. However, we have checks to ensure that this will not happen. As an user, you can decide whether running then in your parsers for helping you during the process, or leave it up to the `normalize()` functions to decide if these are ran or not.
+
+
+<!--
+JMP: I realized we have to make sure that a function has not been used already. Probably we can use the m_cache for this, but most importantly: does this make sense? I.e., do we want to overwrite running public functions when normalize() is ran or do we want to avoid this? Maybe someone used wrongly the function, or do we simply overlook this and ignore it?
```
Comment (JFRudzinski) [inline docs/normalize.md]: I don't think I fully understand why we need to make sure that a public function is only used once. Are there not cases that you want to reuse a base class function multiple times? (I feel like I am missing something in the context here).

File: docs/contact.md [hunk 5]
```diff
@@ -4,17 +4,16 @@ NOMAD is an open source project that warmly welcomes community projects, contrib
...
-    | Dr. José M. Pizarro                   | [jose.pizarro@physik.hu-berlin.de](mailto:jose.pizarro@physik.hu-berlin.de)                 | GW, DMFT, BSE, parsers, workflows, normalizers        |
-    | Dr. Joseph F. Rudzinski (**Coordinator**) | [joseph.rudzinski@physik.hu-berlin.de](mailto:joseph.rudzinski@physik.hu-berlin.de)   | MD, parsers, workflows, normalizers                          |
+| Name | E-mail     | Topics | Github profiles |
+|------|------------|--------|-----------------|
+| Dr. Nathan Daelman | [nathan.daelman@physik.hu-berlin.de](mailto:nathan.daelman@physik.hu-berlin.de) | DFT | [@ndaelman-hu](https://github.com/ndaelman-hu) |
+| Dr. Bernadette Mohr | [mohrbern@physik.hu-berlin.de](mailto:mohrbern@physik.hu-berlin.de) | MD, FF | [@Bernadette-Mohr](https://github.com/Bernadette-Mohr) |
+| Dr. José M. Pizarro | [jose.pizarro@physik.hu-berlin.de](mailto:jose.pizarro@physik.hu-berlin.de) | GW, DMFT, BSE | [@JosePizarro3](https://github.com/JosePizarro3) |
+| Dr. Joseph F. Rudzinski (**Coordinator**) | [joseph.rudzinski@physik.hu-berlin.de](mailto:joseph.rudzinski@physik.hu-berlin.de) | MD, FF | [@JFRudzinski](https://github.com/JFRudzinski) |
```
Comment (JFRudzinski) [inline docs/contact.md]: Can you change my topics to something like "General" (maybe you have an idea for better wording). I would like to give the impression that I am there for overarching issues / general contact.

File: docs/contact.md [hunk 6]
```diff
@@ -4,17 +4,16 @@ NOMAD is an open source project that warmly welcomes community projects, contrib
...
 
 
 Alternatively, you can also:
 
-- Open an issue in the [general NOMAD Github project](https://github.com/nomad-coe/nomad), or in one of the [sub-projects](https://github.com/nomad-coe/nomad/tree/develop/dependencies/parsers) related with specific parsers. Our Github profile tags are [@ndaelman-hu](https://github.com/ndaelman-hu), [@ladinesa](https://github.com/ladinesa), [@JosePizarro3](https://github.com/JosePizarro3), and [@JFRudzinski](https://github.com/JFRudzinski).
-- Write us in the [NOMAD MatSci forum](https://matsci.org/c/nomad/32). Our tags there are @NateD, @ladinesa, @JosePizarro, and @JFRudzinski.
-- Send an email to [support@nomad-lab.eu](mailto:support@nomad-lab.eu). Please, add in the subject "ATTN - Area C".
+- Open an [**issue**](https://github.com/nomad-coe/nomad-schema-plugin-simulation-data/issues) in the [Github project](https://github.com/nomad-coe/nomad-schema-plugin-simulation-data/), and tag any of us.
```
Comment (JFRudzinski) [inline docs/contact.md:17]: Should we suggest that they join the Discord? I would opt for this before sending an email to support, possibly as a first point of contact. In that case, maybe add our Discord handles to the table.

File: docs/general.md [hunk 7]
```diff
@@ -0,0 +1,106 @@
+# `Simulation` base class
+
+We defined a base class named `Simulation`. You can find its Python schema definition in [src/nomad_simulations/general.py](https://github.com/nomad-coe/nomad-schema-plugin-simulation-data/blob/develop/src/nomad_simulations/general.py). In NOMAD, this section will
```
Comment (JFRudzinski) [inline docs/general.md]: I suggest starting with something like: "In NOMAD, all simulation metadata is defined within the `Simulation` base class." 

Then, I would remove "In NOMAD" in the next few sentences.

Comment (JFRudzinski) [inline docs/general.md]: base class == base section? And if so, is this characterization of simulation as a base class correct? Based on the link below, this implies that Simulation will be found under `data.metainfo.basesections` (People are using "base class" now pretty loosly, so I just want to hear your interpretation of what this means).

File: docs/general.md [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+appear under the `data` section for the _archive_ metadata structure of each _entry_.
+
+In NOMAD, a set of [base sections](https://nomad-lab.eu/prod/v1/staging/docs/howto/customization/base_sections.html) derived from the [Basic Formal Ontology (BFO)](https://basic-formal-ontology.org/) are defined. We used them to define `Simulation` as an [`Activity`](http://purl.obolibrary.org/obo/BFO_0000015). The UML diagram is:
```
Comment (JFRudzinski) [inline docs/general.md]: I think I would opt to rephrase this slightly, but it depends a little on your answer to my previous question. So, I put this placeholder and will come back later...

File: docs/general.md [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+
+<div class="click-zoom">
+    <label>
+        <input type="checkbox">
+        <img src="../assets/simulation_base.png" alt="Simulation base section diagram." width="80%" title="Click to zoom in">
+    </label>
+</div>
+
+In fact, `Simulation` inherits from a further abstract concept, `BaseSimulation`. This class contains the general information about the `Program` used, as well as general times of the simulation, e.g., the datetime at which it started (`datetime`) and ended (`datetime_end`). The detailed UML diagram of quantities and functions defined for `Simulation` is thus:
```
Comment (JFRudzinski) [inline docs/general.md]: concept or class?

File: docs/general.md [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+4. `Outputs`: contains all the output properties, as well as references to the `ModelSystem` used to obtain such properties. It might also contain information which will populate `ModelSystem` (e.g., atomic occupations, atomic moments, crystal field energies, etc.).
+
+!!! note "Self-consistent steps, SinglePoint entries, and more complex workflows."
+    In NOMAD, we consider the minimal unit for storing the data in the archive (i.e., an *entry*) as any calculation containing all the self-consistent steps of itself. This is what we call, **`SinglePoint`**. Thus, we do not split each self-consistent step in its own individual entry in the NOMAD archive, but rather store them under the same entry in the archive. Any other complex calculation which combines several differentiated self-consistent calculations must be considered a **workflow** (e.g., a `GW` calculation is usually composed of 2 `SinglePoint` entries: the `DFT SinglePoint` self-consistent calculation + the `GW SinglePoint` self-consistent calculation). You can check the [NOMAD simulations workflow schema](https://github.com/nomad-coe/nomad-schema-plugin-simulation-workflow) for more information.
```
Comment (JFRudzinski) [inline docs/general.md]: This is a bit confusing because it is conflating the definition of the workflow section with the contents of an archive. But in practice, these 2 things are considered separately and depend on the situation. In particular, you may have a workflow whose steps are all contained within a single entry. But a singlepoint self-consistent field calculation does not contain a workflow section.

**First** attempt at rewrite:
"""
The minimal unit for storing data in the NOMAD archive is an [*Entry*](<link to gen docs glossary?>). In the context of simulation data, an entry may contain data from a calculation on an individual system configuration (e.g., a single-point DFT calculation) using **only** the above-mentioned sections of the `Simulation` class. Information from self-consistent iterations to converge properties for this configuration are also contained within these sections.

More complex calculations that involve multiple configurations require the definition of a [**Workflow**](<?>) section within the archive. Depending on the situation, the information from individual workflow steps may be stored within a single or multiple entries. For example, for efficiency, the data from workflows involving a large amount of configurations, e.g., molecular dynamics trajectories, are stored within a single entry. Other standard workflows store the single-point data in separate entries, e.g.,  a `GW` calculation is usually composed of a `DFT SinglePoint` entry and a `GW SinglePoint` entry. Higher-level workflows, which simply connect a series of standard or custom workflows, are typically stored as a separate entry. You can check the [NOMAD simulations workflow schema](https://github.com/nomad-coe/nomad-schema-plugin-simulation-workflow) for more information.
"""

File: docs/general.md [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+
+The simplified schematics for a `Simulation` data section will then be (note that the arrows here are a simple way of visually defining _inputs_ and _outputs_):
```
Comment (JFRudzinski) [inline docs/general.md]: The following schematic represents a simplified representation of the `Simulation` section (note...):

File: docs/general.md [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+<div class="click-zoom">
+    <label>
+        <input type="checkbox">
+        <img src="../assets/simulation_composition.png" alt="Simulation composition diagram." width="90%" title="Click to zoom in">
+    </label>
+</div>
+
+### `Program` {#program}
+
+The `Program` base class section contains all the information about the program / software / code used to perform the simulation. We considered to be a [`(Continuant) Entity`](http://purl.obolibrary.org/obo/BFO_0000002). It is contained within `BaseSimulation` as a sub-section.. The detailed UML diagram is:
```
Comment (JFRudzinski) [inline docs/general.md]: "We consider it to be a ..., contained as a sub-section within `BaseSimulation`"

File: docs/general.md [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+
+<div class="click-zoom">
+    <label>
+        <input type="checkbox">
+        <img src="../assets/program.png" alt="Program quantities and functions UML diagram." width="75%" title="Click to zoom in">
+    </label>
+</div>
+
+
+When [writing a parser](https://nomad-lab.eu/prod/v1/staging/docs/howto/customization/parsers.html), we recommend to start by instantiating the `Program` section and populating its quantities to get acquaintant and learn how to use the NOMAD parsing infrastructure.
```
Comment (JFRudzinski) [inline docs/general.md]: "...its quantities, in order to get acquainted with the NOMAD parsing infrastructure" (it's a bit redundant with both acquainted and learn)

File: docs/general.md [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+
+For example, imagine we have a file, `output_file.txt`, which we want to parse and with the following information:
+```txt
+! * * * * * * *
+! Welcome to SUPERCODE, version 7.0
+...
+```
+
+Then, we can parse the program `name` and `version` by matching the texts (see, e.g., [Wikipedia page for Regular expressions, also called _regex_](https://en.wikipedia.org/wiki/Regular_expression)):
```
Comment (JFRudzinski) [inline docs/general.md]: ~~Then,~~ We can...

File: docs/index.md [hunk 15]
```diff
@@ -3,17 +3,19 @@
...
+NOMAD is a free open-source data management platform for Materials Science which follows the F.A.I.R. (Findable, Accessible, Interoperable, and Reusable) principles. This documentation page is a part of the more [general NOMAD documentation](https://nomad-lab.eu/prod/v1/staging/docs/), as well as on the usage of [NOMAD base sections](https://nomad-lab.eu/prod/v1/staging/docs/howto/customization/base_sections.html).
 
+When designing the sections, we follow [SOLID principles](https://www.geeksforgeeks.org/solid-principle-in-programming-understand-with-real-life-examples/) for object-oriented programming. And throughout this documentation, we will use [UML diagrams](https://en.wikipedia.org/wiki/Class_diagram), both in a simplified and in a detailed manner, to draw the schemas relationships.
 
 
 !!! info "Main contributors"
```
Comment (JFRudzinski) [inline docs/index.md]: This seems a bit unnecessary with the contact page. Maybe you can just add a link or button to there?

File: docs/normalize.md [hunk 16]
```diff
@@ -0,0 +1,66 @@
+# The `normalize()` function
+
+Each base class defined using the NOMAD schema has a set of public functions which can be used at any moment when reading and parsing files in NOMAD (see their corresponding documentation pages). The only special situation happens for the `normalize(archive, logger)` function.
```
Comment (JFRudzinski) [inline docs/normalize.md]: I wouldn't say "see their corresponding..." unless you are going to link something.

"The only special situation..." --> "The `normalize()` function is a special case of such a function, which warrants an in-depth description"

File: docs/normalize.md [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+
+This function is ran within the NOMAD infrastructure by the [`MetainfoNormalizer`](https://github.com/nomad-coe/nomad/blob/develop/nomad/normalizing/metainfo.py) in the following order:
```
Comment (JFRudzinski) [inline docs/normalize.md]: "is run"

File: docs/normalize.md [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+
+1. The deepest child class `normalize()` function is ran before their parents' `normalize()` function.
```
Comment (JFRudzinski) [inline docs/normalize.md]: "A child class' ... is run before their/its parents' ..."

File: docs/normalize.md [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+2. For sibling sections, the `normalize()` function is executed from the smaller to the larger `normalizer_level` attribute. If `normalizer_level` is not set up, the order is established by the attributes definition order in the parent class.
```
Comment (JFRudzinski) [inline docs/normalize.md]: "not set up" --> "not set"?

Also, what if they are set to be the same, I guess then it defaults again to the order in the parent class?

File: docs/normalize.md [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+    sub_section_1 = SubSection(Section1.m_def, repeats=False)
+
+    sub_section_2 = SubSection(Section2.m_def, repeats=True)
+
+    def normalize(self, achive, logger):
+        super().normalize(archive, logger)
+        # Some operations here or before `super().normalize(archive, logger)`
+```
+
+Now, `MetainfoNormalizer` will be ran on the `ParentSection`. Applying **rule 1**, the `normalize()` functions of the `ParentSection`'s childs are executed first. The order of these functions is established by **rule 2** with the `normalizer_level` atrribute, i.e., all the `Section2` (note that `sub_section_2` is a list of sections) `normalize()` functions are ran first, then `Section1.normalize()`. Then, the order of execution will be:
```
Comment (JFRudzinski) [inline docs/normalize.md]: "will be run" and "functions are run first" (The 2nd one is tricky, but I think that is correct)

File: docs/normalize.md [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+
+1. `Section2.normalize()`
+2. `Section1.normalize()`
+3. `ParentSection.normalize()`
+
+In case we do not assign a value to `Section1.normalizer_level` and `Section2.normalizer_level`, all `Section1.normalize()` will run first and then `Section2.normalize()` due to the order of `SubSection` attributes in `ParentSection`. Thus the order will be in this case:
```
Comment (JFRudzinski) [inline docs/normalize.md]: "~~all~~ `Section1.normalize()` will run ~~and then~~ before `Section2.normalize`, due to the ..."

File: docs/normalize.md [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+
+1. `Section1.normalize()`
+2. `Section2.normalize()`
+3. `ParentSection.normalize()`
+
+By checking on the `normalize()` functions and **rule 3**, we can establish whether `ArchiveSection.normalize()` will be ran or not. In `Section1.normalize()`, it will not, while in the other sections, `Section2` and `ParentSection`, it will.
```
Comment (JFRudzinski) [inline docs/normalize.md]: "will be run"

File: docs/normalize.md [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+
+
+### Avoiding double execution of public functions
+
+When the `MetainfoNormalize` is ran, and the `normalize()` functions are executed, these usually have calls for calling private and public functions in the base classes which might have been used already during parsing the files. However, we have checks to ensure that this will not happen. As an user, you can decide whether running then in your parsers for helping you during the process, or leave it up to the `normalize()` functions to decide if these are ran or not.
```
Comment (JFRudzinski) [inline docs/normalize.md]: "is run"

Comment (JFRudzinski) [reply docs/normalize.md]: "as a user"

Comment (JFRudzinski) [reply docs/normalize.md]: "whether running **them**"

Comment (JFRudzinski) [reply docs/normalize.md]: "these usually call private and..."?

Comment (JFRudzinski) [reply docs/normalize.md]: "As a user, you can decide whether to run them directly in within your parser or to leave it up to the `normalize()` function"?

Something like this, but I don't yet understand exactly what the point is...

**Review by JFRudzinski — COMMENTED** [review]
I'll come back to expand upon some things after you respond to a few of my comments

File: docs/contact.md (hunk 6 above)
Comment (JosePizarro3) [reply docs/contact.md:17]: Done, tho Discord recognizes our names, so I wouldn't add our handles

File: docs/general.md (hunk 7 above)
Comment (JosePizarro3) [reply docs/general.md]: Ok, let me rephrase these 3 initial paragraphs and fix the first image.

File: docs/general.md (hunk 9 above)
Comment (JosePizarro3) [reply docs/general.md]: Class is more proper, yes. But let me rephrase a bit these 3 first paragraphs.

File: docs/contact.md [hunk 24]
```diff
@@ -4,17 +4,16 @@ NOMAD is an open source project that warmly welcomes community projects, contrib
...
-!!! info "Main contributors"
-    | Name | E-mail     | Topics |
-    |------|------------|--------|
-    | Dr. Nathan Daelman                    | [nathan.daelman@physik.hu-berlin.de](mailto:nathan.daelman@physik.hu-berlin.de)       | DFT, parsers, normalizers  |
-    | Dr. Alvin Noe Ladines                 | [alvin.noe.ladines@physik.hu-berlin.de](mailto:alvin.noe.ladines@physik.hu-berlin.de) | Parsers, workflows           |
-    | Dr. José M. Pizarro                   | [jose.pizarro@physik.hu-berlin.de](mailto:jose.pizarro@physik.hu-berlin.de)                 | GW, DMFT, BSE, parsers, workflows, normalizers        |
-    | Dr. Joseph F. Rudzinski (**Coordinator**) | [joseph.rudzinski@physik.hu-berlin.de](mailto:joseph.rudzinski@physik.hu-berlin.de)   | MD, parsers, workflows, normalizers                          |
+| Name | E-mail     | Topics | Github profiles |
+|------|------------|--------|-----------------|
+| Dr. Nathan Daelman | [nathan.daelman@physik.hu-berlin.de](mailto:nathan.daelman@physik.hu-berlin.de) | DFT | [@ndaelman-hu](https://github.com/ndaelman-hu) |
```
Comment (ndaelman-hu) [inline docs/contact.md]: Electronic structure, basis sets?

File: docs/contact.md
Comment (ndaelman-hu) [inline docs/contact.md:1]: Typically, contact information, goes into the README, which is shown at the front of the Github page.

**Comment by ndaelman-hu** [comment]
> I want to use this docu page to open more the discussion with PIs in Area C (so that they can actually get involved and give feedback), as I am sure they will be happy to give feedback to the documentation page. I have the feeling this could evolved into an ontology very easily, but for this @ndaelman-hu can give his thoughts.

Yes, this would be very helpful for understanding some of the design choices. Sometimes, it's hard to read these from the code alone.

Regarding the ontology, going off your diagrams, most of them are UML class diagrams, which already have a well-defined semantics. The exception here is the program scheme, but there the arrows are unspecified.

Btw, I wouldn't target top-level ontologies, as designing them probably goes over our heads, and their space is already saturated.
I think the field is more so in dire need of domain-specific ontologies. Our schema could form the basis of such an ontology, yes. In the end, this is just a well-defined nomenclature. Getting the community to adopt it is the tricky part. If they did, it would help tremendously with parsing, though.

File: docs/assets/program.png
Comment (ndaelman-hu) [inline docs/assets/program.png:1]: I would state the legend as:
- "inherits from"
- "is contained in" + specify the meaning of the label

File: docs/assets/simulation_base.png
Comment (ndaelman-hu) [inline docs/assets/simulation_base.png:1]: Rather than inheritance trees, I'd prefer seeing the interaction modes between objects.

File: docs/assets/simulation_composition.png
Comment (ndaelman-hu) [inline docs/assets/simulation_composition.png:1]: label the arrows

**Review by ndaelman-hu — COMMENTED** [review]
short feedback for now

File: docs/assets/simulation_base.png
Comment (JosePizarro3) [reply docs/assets/simulation_base.png:1]: Not sure what you mean.

File: docs/assets/simulation_composition.png
Comment (JosePizarro3) [reply docs/assets/simulation_composition.png:1]: What do you mean?

File: docs/contact.md (hunk 24 above)
Comment (JosePizarro3) [reply docs/contact.md]: I think DFT is pretty good. No need to specify to such level of detail in here.

File: docs/general.md (hunk 10 above)
Comment (JosePizarro3) [reply docs/general.md]: Yeah, I like it, thanks a lot, I was struggling finding how to explain it easily. I incorporate it to the note.

File: docs/normalize.md (hunk 19 above)
Comment (JosePizarro3) [reply docs/normalize.md]: > Also, what if they are set to be the same, I guess then it defaults again to the order in the parent class?

Yes, I added "If `normalizer_level` is not set or if they are the same for two different sections, the order..."

File: docs/normalize.md (hunk 22 above)
Comment (JosePizarro3) [reply docs/normalize.md]: yeah, at some point you would imagine that I know that, but it is obvious that it is not the case 🤣

File: docs/normalize.md (hunk 23 above)
Comment (JosePizarro3) [reply docs/normalize.md]: Yes, I was thinking that the current implementation would be more optimized if we have someway of controlling if a function is executed or not to avoid double executions in the `normalize` function, but I think we don't want that, we anyways want the normalization to run.

File: docs/normalize.md (hunk 4 above)
Comment (JosePizarro3) [reply docs/normalize.md]: Check your last comment on this first review and my answer (I am deleting this section). I'll resolve this.

File: docs/contact.md
Comment (JosePizarro3) [reply docs/contact.md:1]: Ok, I added it, but I also kept it in the documentation page.

File: docs/assets/program.png
Comment (JosePizarro3) [reply docs/assets/program.png:1]: I think the labels are understandable enough. I will add the description of the of the label when containment, tho.

**Comment by JosePizarro3** [comment]
Ok, I think I can merge this and keep working on a second and third parts. 

Please, do a quick check so everything looks reasonable :)

File: docs/contact.md (hunk 24 above)
Comment (ndaelman-hu) [reply docs/contact.md]: People should be able to contact someone regarding the basis sets, band structure and DOS. Who isn't made apparent right now. Add it to my info, pls.

File: docs/assets/simulation_base.png
Comment (ndaelman-hu) [reply docs/assets/simulation_base.png:1]: _class inheritance_ is used as a way for extending user interfaces and functionality. I don't get from this figure what or why it is being extended. the figure gives some overview, but it'll get old soon, and I can just trace the inheritance with the IDE.

As Alan Kay put it, the essence of OOP is _the concept of message passing_ (I guess that in physics, we'd rather say _how 2 objects interact_). Class UML diagrams try to convey this too.

File: docs/assets/simulation_composition.png
Comment (ndaelman-hu) [reply docs/assets/simulation_composition.png:1]: The arrows denote some kind of connection. Which is ..?
I take it they all have the same semantics, though, since the same arrow format is used.

You can also add it as a legend.

File: docs/assets/simulation_base.png
Comment (JosePizarro3) [reply docs/assets/simulation_base.png:1]: I still don't understand what you are trying to say. I have no idea of who Alan Kay is.

This figure is kind of a simplified diagram of the inheritance tree. At this point, we don't want to show all the set of quantities, defs, and methods.

File: docs/assets/simulation_composition.png
Comment (JosePizarro3) [reply docs/assets/simulation_composition.png:1]: This is a simple input/output arrow, it is explained in the text.

**Comment by JosePizarro3** [comment]
@JFRudzinski when you can, do you mind checking again the initial paragraphs in `general.md`? Let me know if now is more clear.

**Comment by JFRudzinski** [comment]
> @JFRudzinski when you can, do you mind checking again the initial paragraphs in `general.md`? Let me know if now is more clear.

Yeah, this looks better :+1:

**Comment by JosePizarro3** [comment]
Cool, I merge then.

Thanks for the reviews, @ndaelman-hu @JFRudzinski ! 👍🏻
